### PR TITLE
Fix race removing threads.

### PR DESCRIPTION
**Pull request by ojanvafai**
setState is async, so it would sometime be operating on a stale
threadIndex. useReducer fixes the race.

### CODE REVIEW

File: ios/TeaMail.xcodeproj/xcshareddata/xcschemes/TeaMail Debug.xcscheme [hunk 1]
```diff
@@ -0,0 +1,78 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (tdresser) [inline ios/TeaMail.xcodeproj/xcshareddata/xcschemes/TeaMail Debug.xcscheme:1]: Ignoring.

File: src/ui/App.tsx [hunk 2]
```diff
@@ -19,9 +19,34 @@ export interface ThreadActions {
...
+
+export interface UpdateThreadListAction {
+  removeThreadId?: string;
+  threads?: gapi.client.gmail.Thread[];
+}
+
 function App(): JSX.Element {
-  const [threadIndex, setThreadIndex] = useState(0);
-  const [threads, setThreads] = useState([] as gapi.client.gmail.Thread[]);
+  function reducer(
```
Comment (tdresser) [inline src/ui/App.tsx:32]: Is it worth verifying that only one of removeThreadId or threads is populated?

Comment (tdresser) [reply src/ui/App.tsx:32]: Or is there a case where it's reasonable to populate both?

File: src/ui/App.tsx [hunk 3]
```diff
@@ -103,29 +125,25 @@ function App(): JSX.Element {
...
+  let numCardsRendered = 10;
+  while (numCardsRendered-- && numCardsRendered < threads.length) {
+    const threadId = threads[numCardsRendered].id as string;
     cards.push(
       <Card
         key={threadId}
         threadId={threadId}
         actions={threadActions}
-        cardSwipedAway={incrementThreadIndex}
+        cardSwipedAway={updateThreadListState}
```
Comment (tdresser) [inline src/ui/App.tsx:142]: I'm confused here - what parameter gets passed to updateThreadListState?

Comment (tdresser) [reply src/ui/App.tsx:142]: I now understand how this works (this is a prop, and the parameter is passed from within the Card) but this still feels a bit confusing.

I think I'd find it a bit clearer if we called this prop "updateThreadListState" or similar?

File: src/ui/App.tsx (hunk 2 above)
Comment (ojanvafai) [reply src/ui/App.tsx:32]: Currently, there's no case the populates both. A problem with post-facto code reviews is addressing small things like this is possibly more work than it's worth. Not sure.

File: src/ui/App.tsx (hunk 3 above)
Comment (ojanvafai) [reply src/ui/App.tsx:142]: Agree it's hard to follow, but from the Card's perspective it doesn't need to know anything about thread list state. Maybe rename it to onRemoveThread?
